### reading_files

#### Reading Files in Python: Axxessing Stored Data

### What is file reading?

_file reading means opening a file and getting its content into python_

#### A file may contain:

* patient notes
* lab reports
* CSV data
* logs
* exported reports

_python can only process file data after it is read into memory_

### Why reding files matters?
_in health informatics, data is often stored in:_

* .txt
* .csv
* .json
* .log
* exported system files

#### examples:
* reading lab test records
* loading patient lists
* analyzing appointment logs
* reading discharge appointment logs


In [1]:
# Opening a file

file = open("clinic_visits.csv","r")

* "clinic_visits.csv" is the file name
* "r" means read mode
* open() gives a file object

_if the file does not exist, Python raises an error_

In [2]:
with open("clinic_visits.csv","r") as file:
    print(file.readline())  # header
    print(file.readline())  # first line of data

visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted

1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No



In [3]:
# read the whole file

file = open("clinic_visits.csv","r")
content = file.read()
print(content)
file.close()

visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted
1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No
1002,P002,67,Male,Cardiology,Arrhythmia,Follow-up,42,220,Yes
1003,P003,15,Female,Pediatrics,Asthma,New,25,60,No
1004,P004,52,Male,Orthopedics,Back Pain,New,30,150,No
1005,P005,73,Female,Cardiology,Heart Failure,Follow-up,55,500,Yes
1006,P006,28,Female,Dermatology,Eczema,New,12,45,No
1007,P007,61,Male,General Medicine,Diabetes,Follow-up,20,95,No
1008,P008,9,Male,Pediatrics,Fever,New,35,50,No
1009,P009,46,Female,Neurology,Migraine,New,27,180,No
1010,P010,58,Male,Orthopedics,Arthritis,Follow-up,40,210,Yes
1011,P011,39,Female,General Medicine,Anemia,New,16,70,No
1012,P012,82,Female,Cardiology,Chest Pain,New,48,320,Yes
1013,P013,22,Male,Dermatology,Acne,Follow-up,10,55,No
1014,P014,65,Female,Neurology,Stroke Review,Follow-up,14,260,No
1015,P015,31,Male,General Medicine,Flu,New,10,65,No



* reads the entire files as one big string
* useful for small files
* not ideal for large clinical datasets

In [4]:
# Read one line at a time

file = open("clinic_visits.csv","r")
line1 = file.readline()
line2 = file.readline()

print("header:", line1)
print("first line of data:", line2)

file.close()

header: visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted

first line of data: 1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No



In [5]:
# Read all lines as a list

file = open("clinic_visits.csv","r")
lines = file.readlines()
print(lines)
file.close()

['visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted\n', '1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No\n', '1002,P002,67,Male,Cardiology,Arrhythmia,Follow-up,42,220,Yes\n', '1003,P003,15,Female,Pediatrics,Asthma,New,25,60,No\n', '1004,P004,52,Male,Orthopedics,Back Pain,New,30,150,No\n', '1005,P005,73,Female,Cardiology,Heart Failure,Follow-up,55,500,Yes\n', '1006,P006,28,Female,Dermatology,Eczema,New,12,45,No\n', '1007,P007,61,Male,General Medicine,Diabetes,Follow-up,20,95,No\n', '1008,P008,9,Male,Pediatrics,Fever,New,35,50,No\n', '1009,P009,46,Female,Neurology,Migraine,New,27,180,No\n', '1010,P010,58,Male,Orthopedics,Arthritis,Follow-up,40,210,Yes\n', '1011,P011,39,Female,General Medicine,Anemia,New,16,70,No\n', '1012,P012,82,Female,Cardiology,Chest Pain,New,48,320,Yes\n', '1013,P013,22,Male,Dermatology,Acne,Follow-up,10,55,No\n', '1014,P014,65,Female,Neurology,Stroke Review,Follow-up,14,260,No\n', '1015,P015,31,Male,General

In [6]:
# loop through the file line by line

file = open("clinic_visits.csv","r")

for line in file:
    print(line)

file.close()

visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted

1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No

1002,P002,67,Male,Cardiology,Arrhythmia,Follow-up,42,220,Yes

1003,P003,15,Female,Pediatrics,Asthma,New,25,60,No

1004,P004,52,Male,Orthopedics,Back Pain,New,30,150,No

1005,P005,73,Female,Cardiology,Heart Failure,Follow-up,55,500,Yes

1006,P006,28,Female,Dermatology,Eczema,New,12,45,No

1007,P007,61,Male,General Medicine,Diabetes,Follow-up,20,95,No

1008,P008,9,Male,Pediatrics,Fever,New,35,50,No

1009,P009,46,Female,Neurology,Migraine,New,27,180,No

1010,P010,58,Male,Orthopedics,Arthritis,Follow-up,40,210,Yes

1011,P011,39,Female,General Medicine,Anemia,New,16,70,No

1012,P012,82,Female,Cardiology,Chest Pain,New,48,320,Yes

1013,P013,22,Male,Dermatology,Acne,Follow-up,10,55,No

1014,P014,65,Female,Neurology,Stroke Review,Follow-up,14,260,No

1015,P015,31,Male,General Medicine,Flu,New,10,65,No



In [7]:
# split header and rows

file = open("clinic_visits.csv","r")
lines = file.readlines()

header = lines[0].strip()
print("header:", header)

first_row = lines[1].strip()
print("first row:", first_row)

header: visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted
first row: 1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No


In [8]:
# split csv row into values

row = lines[1].strip().split(",")
print("row values:", row)

row values: ['1001', 'P001', '34', 'Female', 'General Medicine', 'Hypertension', 'New', '18', '80', 'No']


### Health Informatics Relevance

_Reading files helps you:_

* inspect clinic visit records
* check if the dataset loaded correctly
* read exported EHR or hospital logs
* prepare for analysis of wait times, cost, and admissions

### Exercises

1. Open a file and print only the first line.
2. Read the line by line and print each row without extra newline space.
3. Store all lines in a list and print the number of rows.
4. Split the first data row into columns usinf split(",").

### writing_files

#### Writing Files in Python: Saving Clinic Data and Reports

### What is file writing?

Writing means saving data is from Python into a file.

_In healthcare, this is useful when you want to save:_

* report summaries
* cleaned patient data
* analysis output
* audit notes
* processed records

In [9]:
# write mode

file = open("clinic_report.txt","w")

* "w" means write mode
* if the file already exists, it gets overwritten
if it does not exist, python creates it

_So yes, one small typo can erase your file_

In [ ]:
# writing a simple text file

file = open("clinic_report.txt","w")
file.write("Clinic Visits Report\n")
file.write("Generated from clinic_visits.csv\n")

file.close()

In [10]:
# write a report from the csv file

file = open("clinic_visits.csv","r")
lines = file.readlines()
file.close()

num_records  = len(lines) - 1  # subtract header

report = open("clinic_report.txt", "w")
report.write("Clinic Visits Report\n")
report.write("====================\n")
report.write(f"Total number of visits:{num_records}\n")
report.write(f"Generated from clinic_visits.csv\n")
report.close()

In [11]:
# write selected rows into another file: Suppose we only want to save only admitted patients

with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()


with open("clinic_report.txt", "w") as report:
    report.write("Admitted Patients\n")
    report.write("=================\n")
    for line in lines[1:]:  # Skip header
        row = line.strip().split(",")
        admitted = row[9]

        if admitted == "Yes":
            report.write(line)

In [12]:
# Write formatted patient information

patient_id = "P001"
patient_name = "John Doe"
department = "General Medicine"

with open("clinic_report.txt", "w") as report:
    report.write(f"Patient ID: {patient_id}\n")
    report.write(f"Patient Name: {patient_name}\n")
    report.write(f"Department: {department}\n")

### Health Informatics Relevance

_writing files is useful for:_

* clinic summaries
* patient filtering outputs
* daily system reports
* intermediate analysis files
* preserving results for team review

### Exercises

1. write a text file that says how many rows are in your file.
2. Create a text file and write small summary
3. write only the records into a new file.
4. write a foematted report with two features in your dataset.

### append_mode

#### Append Mode in Python: Adding New clinical records without erasing old ones

_Append mode adds new content to the end of a file_

"w" overwrites existing content
"a" keeps odd content and adds new content

In [16]:
#append mode
file = open("clinic_report.txt", "a")

In [14]:
# append one line

with open("clinic_report.txt", "a") as report:
    report.write("This line is appended to the report.\n")

In [17]:
# append a record from the csv file
with open("clinic_visits.csv", "r") as report:
    lines = report.readlines()

new_log = lines[1].strip()   # first patient record

with open("clinic_report.txt", "a") as report:
    report.write(new_log + "\n")

In [18]:
# append multiple records from the csv file
with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()

with open("clinic_report.txt", "a") as report:
    for line in lines[1:6]:  # append first 5 patient records
        row = line.strip().split(",")
        admitted = row[9] # check if the patient was admitted
        if admitted == "Yes":
            report.write(line)

In [19]:
# append a custom note

with open("clinic_report.txt", "a") as file:
    file.write("2026-06-02: Dataset reviewed for admission trend.\n")

### Health Informatics Relevance

_Append mode is useful for:_

* daily patient logs
* nurse notes
* audit trails
* system events
* incoming visit records

### Exercises
1. Create a log file and keep adding one line every time you run the code.
2. Append all admitted patients from your dataset to txt file
3. Append only department name for each row into a log file.
4. Compare the results of "w" and"a" on the same file.

### with_context_manager

### Using 'with' in Python: Safer file Handling for Clinic Data

In [20]:
# When you manually open a file, you must close it yourself.

file = open("clinic_visits.csv", "r")
# do some processing with the file 

content = file.read() # read the entire file content into a variable
file.close() # close the file after processing.
# If an error occurs before file.close(), the file may remain open, which can lead to resource leaks. Using 'with' statement is recommended to ensure proper file handling.

In [21]:
# append safely using 'with' statement

with open("clinic_report.txt", "a") as file:
    file.write("This is a safe way to append to the file.\n")

### why this is the best practice

_using 'with' is cleaner because:_

* it closes files automatically
* it reduces errors
* it makes code easier to read
* it is the standard python style

### Health Informatics Relevance

_in health systems, file safety matters because files may contain:_

* patient data
* clinic reports
* sensitive records
* audit logs

A bad file operation can corrupt output or lose work. 

### Exercises

* Read your dataset using 'with'.
* write a summary text file using 'with'
* append a line to a log file using 'with'
* Rewrite manual open-close code into a 'with' block.

### csv_manual_python

### Reading and Writing CSV files Manually in Python

_CSV means Comma-Separated Values._

Usually a csv file looks like this:

visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted
1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No
1002,P002,67,Male,Cardiology,Diabetes Follow-up,Follow-up,42,220,Yes
...

_CSV is used because it is simple and works well for sharing tabular data._

In [22]:
# Read all lines from the file

with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()

print(lines[0]) # print the header line
print(lines[1]) # print the first data line


visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted

1001,P001,34,Female,General Medicine,Hypertension,New,18,80,No



In [23]:
# separate header and data lines

with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()

header = lines[0].strip().split(",")
print("Header:", header)

Header: ['visit_id', 'patient_id', 'age', 'gender', 'department', 'diagnosis', 'visit_type', 'wait_time_min', 'cost_usd', 'admitted']


In [24]:
# convert each row into a list

for line in lines[1:]:
    row = line.strip().split(",")
    print("Data Row:", row)

Data Row: ['1001', 'P001', '34', 'Female', 'General Medicine', 'Hypertension', 'New', '18', '80', 'No']
Data Row: ['1002', 'P002', '67', 'Male', 'Cardiology', 'Arrhythmia', 'Follow-up', '42', '220', 'Yes']
Data Row: ['1003', 'P003', '15', 'Female', 'Pediatrics', 'Asthma', 'New', '25', '60', 'No']
Data Row: ['1004', 'P004', '52', 'Male', 'Orthopedics', 'Back Pain', 'New', '30', '150', 'No']
Data Row: ['1005', 'P005', '73', 'Female', 'Cardiology', 'Heart Failure', 'Follow-up', '55', '500', 'Yes']
Data Row: ['1006', 'P006', '28', 'Female', 'Dermatology', 'Eczema', 'New', '12', '45', 'No']
Data Row: ['1007', 'P007', '61', 'Male', 'General Medicine', 'Diabetes', 'Follow-up', '20', '95', 'No']
Data Row: ['1008', 'P008', '9', 'Male', 'Pediatrics', 'Fever', 'New', '35', '50', 'No']
Data Row: ['1009', 'P009', '46', 'Female', 'Neurology', 'Migraine', 'New', '27', '180', 'No']
Data Row: ['1010', 'P010', '58', 'Male', 'Orthopedics', 'Arthritis', 'Follow-up', '40', '210', 'Yes']
Data Row: ['1011', 

In [25]:
# Converting rows into dictionaries

patients = []

with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()

header = lines[0].strip().split(",")

for line in lines[1:]:
    values = line.strip().split(",")
    record = {}

    for i in range(len(header)):
        record[header[i]] = values[i]
    patients.append(record)

print(patients[0]) # print the first patient record as a dictionary


{'visit_id': '1001', 'patient_id': 'P001', 'age': '34', 'gender': 'Female', 'department': 'General Medicine', 'diagnosis': 'Hypertension', 'visit_type': 'New', 'wait_time_min': '18', 'cost_usd': '80', 'admitted': 'No'}


In [28]:
# convert important columns to integers

patients = []
with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()

header = lines[0].strip().split(",")

for line in lines[1:]:
    values = line.strip().split(",")
    record = {}

    for i in range(len(header)):
        record[header[i]] = values[i]

    # Convert age and visit duration to integers
    record["age"] = int(record["age"])
    record["wait_time_min"] = int(record["wait_time_min"])
    record["cost_usd"] = int(record["cost_usd"])

    patients.append(record)

print(patients[0]) # print the first patient record as a dictionary

{'visit_id': '1001', 'patient_id': 'P001', 'age': 34, 'gender': 'Female', 'department': 'General Medicine', 'diagnosis': 'Hypertension', 'visit_type': 'New', 'wait_time_min': 18, 'cost_usd': 80, 'admitted': 'No'}


In [29]:
# find admitted patients manually

with open("clinic_visits.csv", "r") as file:
    lines = file.readlines()

for line in lines[1:]:
    row = line.strip().split(",")
    admitted = row[9]

    if admitted == "Yes":
        print("Admitted patient:", row)

Admitted patient: ['1002', 'P002', '67', 'Male', 'Cardiology', 'Arrhythmia', 'Follow-up', '42', '220', 'Yes']
Admitted patient: ['1005', 'P005', '73', 'Female', 'Cardiology', 'Heart Failure', 'Follow-up', '55', '500', 'Yes']
Admitted patient: ['1010', 'P010', '58', 'Male', 'Orthopedics', 'Arthritis', 'Follow-up', '40', '210', 'Yes']
Admitted patient: ['1012', 'P012', '82', 'Female', 'Cardiology', 'Chest Pain', 'New', '48', '320', 'Yes']


In [30]:
# write a csv file manually

with open("clinic_output.csv", "w") as file:
    file.write("visit_id,patient_id,age,gender,department,diagnosis,visit_type,wait_time_min,cost_usd,admitted\n")
    file.write("2001,P100,40,Female,Cardiology,Checkup,Follow-up,20,150,No\n")
    file.write("2002,P101,55,Male,Neurology,Migraine,New,35,180,Yes\n")

### Health Informatics Relevance

_manual csv handling handling help you:_

* understand how tabular data is stored
* prepare for data cleaning
* filter patient records
* create simple export files
* work with raw data before using pandas

### Exercises

* Read your dataset and print the header seperately
* Covert each row into a dictionary.
* print only the patients whose 'admitted' value is 'Yes'
* save only the 'yes' records into a ew csv file.
* calculate the average 'wait_time_min' manually after reading the csv